src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+          if (gap_minsize_end > 0)
+          {
+            curr =
+                nm->mkNode(MINUS, curr, nm->mkConst(Rational(gap_minsize_end)));
+          }
+          Node ss = nm->mkNode(STRING_SUBSTR, x, curr, lensc);
+          conj.push_back(ss.eqNode(sc));
+        }
+        else if (gap_minsize_end > 0)
+        {
+          Node fit = nm->mkNode(
+              LEQ,
+              nm->mkNode(
+                  PLUS, prev_end, nm->mkConst(Rational(gap_minsize_end))),
+              lenx);
+          conj.push_back(fit);
+        }
+      }
+    }
+    if (success)
+    {
+      // since sep_children is non-empty, conj is non-empty
+      Assert(!conj.empty());
+      Node res = conj.size() == 1 ? conj[0] : nm->mkNode(AND, conj);
+      // process the non-greedy find variables
+      if (!non_greedy_find_vars.empty())
+      {
+        std::vector<Node> children;
+        for (const Node& v : non_greedy_find_vars)
+        {
+          Node bound = nm->mkNode(
+              AND, nm->mkNode(LEQ, d_zero, v), nm->mkNode(LT, v, lenx));
+          children.push_back(bound);
+        }
+        children.push_back(res);
+        Node body = nm->mkNode(AND, children);
+        Node bvl = nm->mkNode(BOUND_VAR_LIST, non_greedy_find_vars);
+        res = nm->mkNode(EXISTS, bvl, body);
+      }
+      // e.g., writing "A" for (str.to.re "A") and _ for re.allchar:
+      //   x in (re.++ "A" (re.* _) "B" (re.* _)) --->
+      //     substr(x,0,1)="A" ^ indexof(x,"B",1)!=-1
+      //   x in (re.++ (re.* _) "A" _ _ _ (re.* _) "B" _ _ (re.* _)) --->
+      //     indexof(x,"A",0)!=-1 ^
+      //     indexof( x, "B", indexof( x, "A", 0 ) + 1 + 3 ) != -1 ^
+      //     indexof( x, "B", indexof( x, "A", 0 ) + 1 + 3 )+1+2 <= len(x)
+      return returnElim(atom, res, "concat-with-gaps");
+    }
+  }
+
+  if (!options::regExpElimAgg())
+  {
+    return Node::null();
+  }
+  // only aggressive rewrites below here
+
+  // if the first or last child is constant string, split
+  Node sStartIndex = d_zero;
+  Node sLength = lenx;
+  std::vector<Node> sConstraints;
+  std::vector<Node> rexpElimChildren;
+  for (unsigned r = 0; r < 2; r++)
+  {
+    unsigned index = r == 0 ? 0 : nchildren - 1;
+    Node c = children[index];
+    if (c.getKind() == STRING_TO_REGEXP)
+    {
+      Node s = c[0];
+      Node lens = nm->mkNode(STRING_LENGTH, s);
+      Node sss = r == 0 ? d_zero : nm->mkNode(MINUS, lenx, lens);
+      Node ss = nm->mkNode(STRING_SUBSTR, x, sss, lens);
+      sConstraints.push_back(ss.eqNode(s));
+      if (r == 0)
+      {
+        sStartIndex = lens;
+      }
+      sLength = nm->mkNode(MINUS, sLength, lens);
+    }
+    if (r == 1 && !sConstraints.empty())
+    {
+      // add the middle children
+      for (unsigned i = 1; i < (nchildren - 1); i++)
+      {
+        rexpElimChildren.push_back(children[i]);
+      }
+    }
+    if (c.getKind() != STRING_TO_REGEXP)
+    {
+      rexpElimChildren.push_back(c);
+    }
+  }
```

Comment:
Can we assert that `rexpElimChildren.size() + sConstraints.size() == nchildren`?